### PR TITLE
EVM version

### LINKED ISSUE #3273: Remove byzantium-only warning
Using any of these assembly instructions:

- CREATE2
- RETURNDATACOPY
- RETURNDATASIZE
- STATICCAL

Will result in the compiler throwing a warning.

It seems like this might be something that should be removed or at least could be silenced.

### PR DESCRIPTION
**Pull request by chriseth**
- [x] tests for metadata

Fixes #3273.

### CODE REVIEW

File: libsolidity/interface/EVMVersion.h [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+ */
+class EVMVersion
+{
+public:
+	EVMVersion() {}
+
+	static EVMVersion homestead() { return {Version::Homestead}; }
+	static EVMVersion byzantium() { return {Version::Byzantium}; }
+
+	static boost::optional<EVMVersion> fromString(std::string const& _version)
```
Comment (axic) [inline libsolidity/interface/EVMVersion.h]: Are we moving over to `boost::optional`? A lot of places started to use it now.

Comment (chriseth) [reply libsolidity/interface/EVMVersion.h]: I just think it is very convenient in this situation.

**Comment by chriseth** [comment]
Ready for review.

**Comment by chriseth** [comment]
@axic ready for review.

**Comment by chriseth** [comment]
Actually I just realized we have a third test combination to run: contracts compiled for homestead, running on a byzantium VM...

File: libsolidity/inlineasm/AsmAnalysis.cpp [hunk 2]
```diff
@@ -533,19 +533,33 @@ void AsmAnalyzer::expectValidType(string const& type, SourceLocation const& _loc
...
-	static set<solidity::Instruction> futureInstructions{
-		solidity::Instruction::CREATE2,
-		solidity::Instruction::RETURNDATACOPY,
-		solidity::Instruction::RETURNDATASIZE,
-		solidity::Instruction::STATICCALL
-	};
-	if (futureInstructions.count(_instr))
+	// We assume that returndatacopy, returndatasize and staticcall are either all available
+	// or all not available.
+	solAssert(m_evmVersion.hasReturndatacopy() == m_evmVersion.hasStaticCall(), "");
```
Comment (axic) [inline libsolidity/inlineasm/AsmAnalysis.cpp]: Does `hasReturndatacopy` also means `hasReturndatasize`?

File: test/RPCSession.cpp [hunk 3]
```diff
@@ -224,6 +231,7 @@ void RPCSession::test_setChainParams(vector<string> const& _accounts)
 			"blockReward": "0x",
 			"allowFutureBlocks": true,
 			"homesteadForkBlock": "0x00",
+			)" + enableByzantium + R"(
 			"EIP150ForkBlock": "0x00",
```
Comment (axic) [inline test/RPCSession.cpp]: This is "tangerine whistle".

File: test/RPCSession.cpp [hunk 4]
```diff
@@ -224,6 +231,7 @@ void RPCSession::test_setChainParams(vector<string> const& _accounts)
...
 			"EIP158ForkBlock": "0x00"
```
Comment (axic) [inline test/RPCSession.cpp]: This is "spurious dragon".

File: test/libsolidity/SolidityABIJSON.cpp [hunk 5]
```diff
@@ -44,6 +44,8 @@ class JSONInterfaceChecker
 	{
 		m_compilerStack.reset(false);
 		m_compilerStack.addSource("", "pragma solidity >=0.0;\n" + _code);
+		m_compilerStack.setEVMVersion(dev::test::Options::get().evmVersion());
+		m_compilerStack.setOptimiserSettings(dev::test::Options::get().optimize);
```
Comment (axic) [inline test/libsolidity/SolidityABIJSON.cpp:48]: This was missed but shouldn't make any difference.

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 6]
```diff
@@ -6857,7 +6857,7 @@ BOOST_AUTO_TEST_CASE(create2_as_variable)
 	)";
 	CHECK_ALLOW_MULTI(text, (std::vector<std::pair<Error::Type, std::string>>{
 		{Error::Type::Warning, "Variable is shadowed in inline assembly by an instruction of the same name"},
-		{Error::Type::Warning, "only available after the Metropolis"},
+		{Error::Type::Warning, "The \"create2\" instruction is not supported for the VM version"},
```
Comment (axic) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp]: "for the"? Shouldn't that be "in the"?

File: libsolidity/inlineasm/AsmAnalysis.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/inlineasm/AsmAnalysis.cpp]: Yes, I can rename it to `hasReturndatacopyAndReturndatasize` or add a comment.

File: test/libsolidity/SolidityNameAndTypeResolution.cpp (hunk 6 above)
Comment (chriseth) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp]: yep!

Comment (chriseth) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp]: by?

File: libsolidity/interface/EVMVersion.h [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+		if (_version == "homestead")
+			return homestead();
+		else if (_version == "byzantium")
+			return byzantium();
+		else
+			return {};
+	}
+
+	bool operator==(EVMVersion const& _other) const { return m_version == _other.m_version; }
+	bool operator!=(EVMVersion const& _other) const { return !this->operator==(_other); }
```
Comment (axic) [inline libsolidity/interface/EVMVersion.h]: Should also have comparison operators if we introduce tangerine whistle now.

File: libsolidity/interface/EVMVersion.h [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+	std::string name() const { return m_version == Version::Byzantium ? "byzantium" : "homestead"; }
+
+	bool hasReturndatacopy() const { return m_version == Version::Byzantium; }
+	bool hasStaticCall() const { return m_version == Version::Byzantium; }
```
Comment (axic) [inline libsolidity/interface/EVMVersion.h]: I'm not sure if we should have such a granularity. One option is this, the other is comparison whether currently version >= byzantium, though that assumes features can't be removed.

File: libsolidity/interface/EVMVersion.h [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+	/// Whether we have to retain the costs for the call opcode itself (false),
+	/// or whether we can just forward easily all remaining gas (true).
+	bool canOverchargeGasForCall() const
+	{
+		// @TODO when exactly was this introduced? Was together with the call stack fix.
```
Comment (axic) [inline libsolidity/interface/EVMVersion.h]: Tangerine Whistle, see https://github.com/ethereum/EIPs/blob/master/EIPS/eip-608.md

File: libsolidity/inlineasm/AsmAnalysis.cpp (hunk 2 above)
Comment (axic) [reply libsolidity/inlineasm/AsmAnalysis.cpp]: I think a comment is enough. Maybe `hasReturndata` was the generic name of the feature?

File: test/libsolidity/SolidityNameAndTypeResolution.cpp (hunk 6 above)
Comment (axic) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp]: Either is fine for me.

File: circle.yml [hunk 10]
```diff
@@ -5,10 +5,6 @@ jobs:
       - image: trzeci/emscripten:sdk-tag-1.37.21-64bit
     steps:
       - checkout
-      - run:
-          name: Init submodules
-          command: |
-            git submodule update --init
```
Comment (axic) [inline circle.yml:11]: Why were these removed?

**Comment by axic** [comment]
> Actually I just realized we have a third test combination to run: contracts compiled for homestead, running on a byzantium VM...

Probably the simplest way is adding another commandline option to the tests for `--runtime-evm-version` which changes the version in the genesis.

Then we could have more combinations on the CI.

**Comment by chriseth** [comment]
Added the other versions.

File: libsolidity/inlineasm/AsmAnalysis.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/inlineasm/AsmAnalysis.cpp]: I will change it to `supportsReturndata`.

File: circle.yml (hunk 10 above)
Comment (chriseth) [reply circle.yml:11]: I think we don't have any submodules anymore, do we? I can remove it from this PR.

**Comment by chriseth** [comment]
Updated. Will leave it at the current test combinations for now.

**Comment by chriseth** [comment]
Updated.

**Comment by axic** [comment]
I'll rebase this after #2541 and fix up the asm analyzer part.

It is only in conflict with `tests.sh` atm.

File: scripts/tests.sh [hunk 11]
```diff
@@ -85,13 +83,31 @@ ETH_PID=$!
...
+        log=--logger=JUNIT,test_suite,$log_directory/opt_$vm.xml
+      else
+        log=--logger=JUNIT,test_suite,$log_directory/noopt_$vm.xml
+      fi
+    fi
+    set +e
+    "$REPO_ROOT"/build/test/soltest --show-progress $log -- "$optimize" --evm-version "$vm" --ipcpath /tmp/test/geth.ipc
+    THIS_ERR=$?
+    set -e
+    if [ $? -ne 0 ]; then ERRO_CODE=$?; fi
```
Comment (axic) [inline scripts/tests.sh]: Typo, the variable is called `ERROR_CODE`

File: scripts/tests.sh [hunk 12]
```diff
@@ -85,13 +83,31 @@ ETH_PID=$!
...
+      if [ -n "$optimize" ]
+      then
+        log=--logger=JUNIT,test_suite,$log_directory/opt_$vm.xml
+      else
+        log=--logger=JUNIT,test_suite,$log_directory/noopt_$vm.xml
+      fi
+    fi
+    set +e
+    "$REPO_ROOT"/build/test/soltest --show-progress $log -- "$optimize" --evm-version "$vm" --ipcpath /tmp/test/geth.ipc
+    THIS_ERR=$?
```
Comment (axic) [inline scripts/tests.sh]: Unused variable?

**Comment by axic** [comment]
Should introduce constaninople too.

**Comment by axic** [comment]
@chriseth please check my rebase of `tests.sh`

File: scripts/tests.sh [hunk 13]
```diff
@@ -116,9 +115,8 @@ do
     fi
     set +e
     "$REPO_ROOT"/build/test/soltest $progress $log -- "$optimize" --evm-version "$vm" --ipcpath /tmp/test/geth.ipc
-    THIS_ERR=$?
     set -e
-    if [ $? -ne 0 ]; then ERROR_CODE=$?; fi
+    if [ $? -ne 0 ]; then exit $?; fi
```
Comment (chriseth) [inline scripts/tests.sh]: Why this complication? Don't we get the same effect by just removing `set +e`?

Comment (chriseth) [reply scripts/tests.sh]: But in general, I agree, we should not continue with the other tests after a failed run.

**Comment by chriseth** [comment]
Added constantinople.

File: libevmasm/Assembly.cpp [hunk 14]
```diff
@@ -353,7 +353,7 @@ void Assembly::injectStart(AssemblyItem const& _i)
 	m_items.insert(m_items.begin(), _i);
 }
 
-Assembly& Assembly::optimise(bool _enable, bool _isCreation, size_t _runs)
+Assembly& Assembly::optimise(bool _enable, EVMVersion _evmVersion, bool _isCreation, size_t _runs)
```
Comment (axic) [inline libevmasm/Assembly.cpp:356]: Woohoo, I should be using this for the bitwise shifting optimiser rules in #3580.

Comment (chriseth) [reply libevmasm/Assembly.cpp:356]: Yes, it was a really painful process. This setting is needed literally everywhere...

File: test/liblll/EndToEndTest.cpp [hunk 15]
```diff
@@ -583,24 +586,34 @@ BOOST_AUTO_TEST_CASE(allgas)
...
-	char const* sourceCode = R"(
-		(returnlll
-			(send 0xdead 42))
-	)";
-	compileAndRun(sourceCode);
-	callFallbackWithValue(42);
-	BOOST_CHECK(balanceAt(Address(0xdead)) == 42);
+	// "send" does not retain enough gas to be able to pay for account creation.
+	// Disabling for non-tangerineWhistle VMs.
+	if (dev::test::Options::get().evmVersion().canOverchargeGasForCall())
```
Comment (axic) [inline test/liblll/EndToEndTest.cpp:591]: Hm, can we output a boost message that these are skipped due to the vm?

Also, how about?
```
if (..)
  return;
```

File: test/liblll/EndToEndTest.cpp [hunk 16]
```diff
@@ -583,24 +586,34 @@ BOOST_AUTO_TEST_CASE(allgas)
...
+	{
+		char const* sourceCode = R"(
+			(returnlll
+				(send 0xdead 42))
```
Comment (axic) [inline test/liblll/EndToEndTest.cpp:595]: Eventually `send` and `msg` should be updated I think.

@benjaminion @zigguratt what do you think?

File: test/liblll/EndToEndTest.cpp (hunk 15 above)
Comment (chriseth) [reply test/liblll/EndToEndTest.cpp:591]: I tried return, but the macro seems to create a function that returns something.

Comment (chriseth) [reply test/liblll/EndToEndTest.cpp:591]: I'll look into telling boost to skip a test, it seems to have such a feature.

Comment (chriseth) [reply test/liblll/EndToEndTest.cpp:591]: Changed it accordingly. This will probably show up in the xml test report.

**Review by axic — APPROVED** [review]
I think it looks good. I wouldn't have added `evmVersion` to LLL, but it can actually be useful (will address that separately).

**Comment by axic** [comment]
It is failing:
```
Testing soljson via the fuzzer...
--> Running tests using  --evm-version homestead...
Commandline tests successful.
Boost.Test framework internal error: unknown reason
Exited with code 200
```

**Comment by chriseth** [comment]
It seems there is a bug in boost which causes skipped tests to be treated as errors: https://svn.boost.org/trac10/ticket/12095

**Comment by chriseth** [comment]
I will revert to the previous version for now (which just had a `if (..) {...}`. These LLL tests already cost me 1-2 hours, although LLL is out of scope.
